tool/src/c2/header.rs

```diff
@@ -46,54 +39,71 @@ pub struct Header {
     /// Foo make_foo(uint8_t field1, bool field2);
     /// ```
     pub body: String,
+    /// What string to use for indentation.
+    pub indent_str: &'static str,
 }
 
 impl Header {
-    pub fn new(identifier: String) -> Self {
+    pub fn new(path: String) -> Self {
         Header {
-            identifier,
+            path,
             includes: BTreeSet::new(),
-            forwards: BTreeSet::new(),
+            decl_include: None,
             body: String::new(),
+            indent_str: "  ",
         }
     }
 }
 
+impl fmt::Write for Header {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        self.body.write_str(s)
+    }
+    fn write_char(&mut self, c: char) -> fmt::Result {
+        self.body.write_char(c)
+    }
+    fn write_fmt(&mut self, args: fmt::Arguments<'_>) -> fmt::Result {
+        self.body.write_fmt(args)
+    }
+}
+
 impl fmt::Display for Header {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let mut forwards = String::new();
         let mut includes = String::from(BASE_INCLUDES);
         for i in &self.includes {
-            includes += &format!("#include \"{}.h\"\n", i);
+            includes += &format!("#include \"{}\"\n", i);
         }
-        for f in &self.forwards {
-            forwards += &format!("typedef struct {f} {f};\n");
-        }
-        let identifier = &self.identifier;
-        let body = &self.body;
+        let decl_header_include: Cow<str> = match self.decl_include {
+            Some(ref v) => format!("\n#include \"{v}\"\n").into(),
+            None => "".into(),
+        };
+        let header_guard = &self.path;
+        let header_guard = header_guard.replace(".d.h", "_D_H");
+        let header_guard = header_guard.replace(".h", "_H");
+        let body: Cow<str> = if self.body.is_empty() {
+            "// No Content\n\n".into()
+        } else {
+            self.body.replace('\t', self.indent_str).into()
```

Comment:
thought: not quite a fan of this way of handling indents, especially since it means that all of our strings must contain `\t` escapes and they get less readable. But it's not a huge deal.